Timeout query with proper error if worker is down

ShahabT opened the pull request:
## What changed?
<!-- Describe what has changed in this PR -->
Fixing https://github.com/temporalio/temporal/issues/1546 by rejecting Workflow Queries with the following message if no poller has been seen recently: `"no poller seen for task queue recently, worker may be down"`.
We wait for 1 sec less than the actual timeout before returning this error.

## Why?
<!-- Tell your future self why have you made these changes -->
The previous behavior did not provide useful error messages in cases where the workers are down.

## How did you test it?
<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
unit test

## Potential risks
<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
None

## Is hotfix candidate?
<!-- Is this PR a hotfix candidate or does it require a notification to be sent to the broader community? (Yes/No) -->
No

bergundy commented:
I don't think this is the desired behavior.

Users should be able to write a test for their own code that concurrently sends a query and starts a worker.

I think we should convert context deadlines to more informative error messages.

Please make sure that you're returning the proper grpc error code in this situation (I think you can use DEADLINE EXCEEDED with a custom error string).

https://grpc.github.io/grpc/core/md_doc_statuscodes.html

stephanos commented on service/matching/matcher_test.go:
```diff
@@ -359,6 +359,26 @@ func (t *MatcherTestSuite) TestSyncMatchFailure() {
+			task.forwardedFrom = req.GetForwardedSource()
+			resp, err := t.rootMatcher.OfferQuery(ctx, task)
+			t.Nil(resp)
+			t.Assert().Error(err, errNoRecentPoller)
```
I can't quite put together why this test works since the default timeout is 20s?

ShahabT replied:
it was working because there was no pollers at all. Plz note that the 20s is not the timeout, it's the window in which we expect to see a poller.

ShahabT commented:
> I don't think this is the desired behavior.
> 
> Users should be able to write a test for their own code that concurrently sends a query and starts a worker.
> 
> I think we should convert context deadlines to more informative error messages.
> 
> Please make sure that you're returning the proper grpc error code in this situation (I think you can use DEADLINE EXCEEDED with a custom error string).
> 
> https://grpc.github.io/grpc/core/md_doc_statuscodes.html

This a good point @bergundy. I want to hear @yiminc 's opinion since he suggested the idea of fast failing queries.

ShahabT commented:
@bergundy please checkout the new behavior. I think it addresses your concerns. I also added the "DEADLINE EXCEEDED" grpc error code in the returned error.

bergundy commented on common/dynamicconfig/constants.go:
```diff
@@ -470,6 +470,8 @@ const (
 	// MatchingMaxWaitForPollerBeforeFwd in presence of a non-negligible backlog, we resume forwarding tasks if the
 	// duration since last poll exceeds this threshold.
 	MatchingMaxWaitForPollerBeforeFwd = "matching.maxWaitForPollerBeforeFwd"
+	// QueryPollerUnavailableWindow WF Query tasks are immediately rejected if no poller has been seen within the window
```
```suggestion
	// QueryPollerUnavailableWindow WF Queries are rejected after a while if no poller has been seen within the window
```

bergundy commented on service/matching/config.go:
```diff
@@ -240,6 +243,9 @@ func newTaskQueueConfig(id *taskQueueID, config *Config, namespace namespace.Nam
 		MaxWaitForPollerBeforeFwd: func() time.Duration {
 			return config.MaxWaitForPollerBeforeFwd(namespace.String(), taskQueueName, taskType)
 		},
+		QueryPollerUnavailableWindow: func() time.Duration {
+			return config.QueryPollerUnavailableWindow()
+		},
```
```suggestion
		QueryPollerUnavailableWindow: config.QueryPollerUnavailableWindow,
```

bergundy commented on service/matching/matcher.go:
```diff
@@ -79,7 +79,8 @@ const (
 	// Sentinel error to redirect while blocked in matcher.
-	errInterrupted = errors.New("interrupted offer")
+	errInterrupted    = errors.New("interrupted offer")
+	errNoRecentPoller = errors.New("no poller seen for task queue recently, worker may be down")
```
```suggestion
	errNoRecentPoller = status.Error(codes.DeadlineExceeded, "no poller seen for task queue recently, worker may be down")
```

bergundy commented on service/matching/matcher.go:
```diff
@@ -230,6 +231,16 @@ func (tm *TaskMatcher) OfferQuery(ctx context.Context, task *internalTask) (*mat
+	if deadline, ok := ctx.Deadline(); ok && fwdrTokenC == nil {
+		// Create a timeout for 90% time of the actual timeout. This is so that we have an opportunity to customize the
+		// "context deadline exceeded" error when user is querying a workflow without having started the workers.
+		noPollerTimeout := time.Until(deadline) * 90 / 100
```
I would max this with (deadline - 1 second).

bergundy commented on service/matching/handler.go:
```diff
@@ -251,7 +254,11 @@ func (h *Handler) QueryWorkflow(
 
-	return h.engine.QueryWorkflow(ctx, request)
+	resp, err := h.engine.QueryWorkflow(ctx, request)
+	if errors.Is(err, errNoRecentPoller) {
+		return nil, status.Error(codes.DeadlineExceeded, err.Error())
+	}
```
If you take my suggestion from below, you don't need this if statement.

bergundy reviewed (commented):
I think this is fine for the most part but there will likely be cases where we've seen a poller in the configurable window but the query will still time out.

I suggest simplifying and checking on the frontend if the history client request timed out and returning a deadline exceed error with a custom message.
This doesn't work if the client sets a shorter timeout than the default 30 second wait but I imagine most clients don't explicitly set request deadlines, especially not newer users that may get confused by the generic deadline exceeded error.

That being said, I think this is approach is valid and does cover some of the cases.

bergundy commented on service/matching/matcher.go:
```diff
@@ -249,6 +262,13 @@ func (tm *TaskMatcher) OfferQuery(ctx context.Context, task *internalTask) (*mat
+		case <-noPollerCtxC:
+			// only error if there has not been a recent poller. Otherwise, let it wait for the remaining time
+			// hopping for a match, or ultimately returning the default CDE error.
+			if tm.timeSinceLastPoll() > tm.config.QueryPollerUnavailableWindow() {
```
Should we only do this on the root partition?
It's more likely to see pollers due to forwarding.

bergundy commented on service/matching/matcher.go:
```diff
@@ -79,7 +81,8 @@ const (
 	// Sentinel error to redirect while blocked in matcher.
-	errInterrupted = errors.New("interrupted offer")
+	errInterrupted    = errors.New("interrupted offer")
+	errNoRecentPoller = status.Error(codes.DeadlineExceeded, "no poller seen for task queue recently, worker may be down")
```
Now I'm wondering if this should be a failed precondition error since this doesn't wait for the whole deadline to expire.

https://grpc.github.io/grpc/core/md_doc_statuscodes.html
